### DISCUSSION TITLE
How to train Yolo to recognize classes that are not in the training dataset? Plz help

### DISCUSSION THREAD

**Discussion by lexDiller** [ultralytics/ultralytics#7081] [Q&A]
Hello everyone, I recently started in computer vision. There is an enterprise where cameras need to determine whether an employee is wearing personal protective equipment or not. Unfortunately, it is absolutely impossible to assemble a training dataset on these cameras in the data of violators, since everyone follows safety rules. After training, the model will be able to recognize only people in personal protective equipment (helmets, gloves, etc.) How do I train Yolo so that she can recognize violations? From my thoughts, the first is to use isolating forest, but I will have to transform the images into vector space, or somehow work with IoU. Plz suggest any ways to solve this problem.

**Comment by pderrenger** [comment]
@lexDiller to train a YOLO model to recognize classes that are not present in your training dataset, you can consider the following approaches:

1. **Synthetic Data Generation**: Create synthetic images of people without personal protective equipment (PPE) using image editing tools or 3D rendering software. This can help the model learn the appearance of non-compliance cases.

2. **Transfer Learning**: Use a pretrained YOLO model that has been trained on a diverse set of objects and fine-tune it on your dataset with PPE. The model may generalize and detect anomalies (people without PPE) based on the learned features.

3. **Data Augmentation**: Apply data augmentation techniques to your existing dataset to simulate the absence of PPE. For example, you could erase parts of the PPE in images to mimic non-compliance.

4. **Anomaly Detection**: Train your YOLO model to detect PPE accurately. Then, use an anomaly detection algorithm to flag instances where the expected PPE is not detected by the model.

5. **Negative Samples**: If possible, collect a small set of images or use stock photos of people without PPE and include them as negative samples in your training dataset.

6. **Open Set Recognition**: Explore open set recognition techniques that allow the model to identify objects as "unknown" if they don't fit into known classes.

Remember, it's crucial to ensure that your model is not only accurate but also reliable and doesn't produce false positives, which could lead to ignoring actual safety violations. Regular validation and updates to the model with new data as it becomes available will help maintain its performance.

**Reply by lexDiller** [reply]
Thanks a lot! I have already completed this project)

**Reply by pderrenger** [reply]
@lexDiller great to hear that you've completed your project! If you have any more questions in the future or need assistance with another computer vision task, feel free to reach out. Good luck with your future endeavors in computer vision! 😊👍

**Reply by nihad3** [reply]
> Thanks a lot! I have already completed this project)
Hi @lexDiller , I’m facing the same issue I only have data of workers wearing PPE and I can't figure out how  to detect missing PPE (like no helmet or no vest), especially since I need to trigger alerts on a dashboard when there’s a violation.

Could you share how you solved this in your project or which approach you followed?

Thanks!

**Reply by lexDiller** [reply]
Good evening @nihad3! I'm sorry for the long answer. Honestly, I don't remember exactly, but it seems like the problem was solved by enriching the dataset in two ways:
1) Generation of synthetic data. At that time, it seemed like even SOTA was not producing reliable generations, but still something could be taken away. Today, there are many models that can actually generate anomalies for a custom domain so well.
2) The second option is much simpler - to agree with the production that for a while a person is sent to the factory with obvious violations to collect the dataset.
I hope I can help you with your question.

**Comment by ArianGhaderi99** [comment]
This is a common problem in PPE/defect detection where “violations” are rare or unavailable.

You actually don’t need Isolation Forest or feature vector space methods for YOLO. Instead, you can approach it in a few practical ways:

### 1. Reframe the problem (recommended)
Instead of detecting “violation”, train YOLO to detect PPE items directly:
- helmet
- gloves
- vest, etc.

Then define logic rules:
- If person detected AND helmet not detected → violation
- If person detected AND gloves missing → violation

This avoids needing violation images completely.

### 2. Use “negative samples”
Even if you cannot collect real violations, you can:
- Use images of people without PPE from public datasets
- Use synthetic removal (see below)

### 3. Synthetic data augmentation
You can artificially create violations by:
- removing helmets/gloves from PPE images (inpainting)
- pasting PPE on non-PPE people (or vice versa)
- using tools like segment-anything + editing

### 4. One-class detection (less recommended)
You can train the model only on “safe PPE” and treat low-confidence / missing detections as anomaly, but this is usually less stable than object detection + rule-based logic.

### Conclusion:
The most robust industrial approach is:
YOLO detects PPE objects → rule engine decides violation.

This is widely used in real-world safety monitoring systems.

**Reply by glenn-jocher** [reply]
A practical way to handle this is exactly as described above: train Ultralytics YOLO to detect `person` plus PPE items, then mark a violation in post-processing when a required item is missing; if you want a ready example, the [Construction-PPE dataset](https://docs.ultralytics.com/datasets/detect/construction-ppe) already includes both worn PPE classes and missing-PPE classes like `no_helmet`, `no_goggle`, `no_gloves`, and `no_boots`, which is a good starting point for this use case.
